lib/Jsrt/Jsrt.cpp

```diff
@@ -2746,3 +3230,504 @@ CHAKRA_API JsRunSerializedScriptWithCallback(_In_ JsSerializedScriptLoadSourceCa
+#endif
+}
+
+CHAKRA_API JsTTDReStartTimeTravelAfterRuntimeOperation()
+{
+#if !ENABLE_TTD
+    return JsErrorCategoryUsage;
+#else
+    JsrtContext *currentContext = JsrtContext::GetCurrent();
+    if(currentContext->GetRuntime()->GetThreadContext()->TTDLog != nullptr)
+    {
+        currentContext->GetRuntime()->GetThreadContext()->TTDLog->PopMode(TTD::TTDMode::ExcludedExecution);
+    }
+
+    return JsNoError;
+#endif
+}
+
+CHAKRA_API JsTTDNotifyHostCallbackCreatedOrCanceled(_In_ bool isCreated, _In_ bool isCancel, _In_ bool isRepeating, _In_ JsValueRef function, _In_ INT64 callbackId)
+{
+#if !ENABLE_TTD
+    return JsErrorCategoryUsage;
+#else
+    JsrtContext *currentContext = JsrtContext::GetCurrent();
+    Js::ScriptContext* scriptContext = currentContext->GetScriptContext();
+
+    if(PERFORM_JSRT_TTD_RECORD_ACTION_CHECK(scriptContext))
+    {
+        Js::JavascriptFunction* jsFunction = (function != nullptr) ? Js::JavascriptFunction::FromVar(function) : nullptr;
+
+        scriptContext->GetThreadContext()->TTDLog->RecordJsRTCallbackOperation(scriptContext, isCreated, isCancel, isRepeating, jsFunction, callbackId);
+    }
+
+    return JsNoError;
+#endif
+}
+
+CHAKRA_API JsTTDNotifyYield()
+{
+#if !ENABLE_TTD
+    return JsErrorCategoryUsage;
+#else
+    JsrtContext *currentContext = JsrtContext::GetCurrent();
+    Js::ScriptContext* scriptContext = currentContext->GetScriptContext();
+
+    if(PERFORM_JSRT_TTD_RECORD_ACTION_CHECK(scriptContext))
+    {
+        scriptContext->GetThreadContext()->TTDLog->RecordJsRTEventLoopYieldPoint(scriptContext);
+    }
+
+    return JsNoError;
+#endif
+}
+
+CHAKRA_API JsTTDPrepContextsForTopLevelEventMove(_In_ JsRuntimeHandle runtimeHandle, _In_ INT64 targetEventTime, _Out_ INT64* targetStartSnapTime)
+{
+#if !ENABLE_TTD_DEBUGGING
+    return JsErrorCategoryUsage;
+#else
+    JsrtContext *currentContext = JsrtContext::GetCurrent();
+    Js::ScriptContext* scriptContext = currentContext->GetScriptContext();
+    if(scriptContext->GetThreadContext()->TTDLog == nullptr)
+    {
+        AssertMsg(false, "Should only happen in TT debugging mode.");
+        return JsErrorCategoryUsage;
+    }
+
+    //Make sure we don't have any pending recorded exceptions
+    if(scriptContext->HasRecordedException())
+    {
+        scriptContext->GetAndClearRecordedException(nullptr);
+    }
+
+    //a special indicator to use the time from the argument flag or log diagnostic report
+    if(targetEventTime == -2)
+    {
+        targetEventTime = scriptContext->GetThreadContext()->TTDLog->GetKthEventTime(Js::Configuration::Global.flags.TTDStartEvent);
+    }
+
+    bool createFreshCtxs = false;
+    *targetStartSnapTime = scriptContext->GetThreadContext()->TTDLog->FindSnapTimeForEventTime(targetEventTime, &createFreshCtxs);
+
+    if(createFreshCtxs)
+    {
+        try
+        {
+            AUTO_NESTED_HANDLED_EXCEPTION_TYPE((ExceptionType)(ExceptionType_OutOfMemory | ExceptionType_StackOverflow));
```

Comment:
Why not use GlobalAPIWrapper?
Are other TTD APIs allowed when heap enum is in progress or we are in collect callback? if not the all TTD APIs should use ContextAPINoScriptWrapper so that CheckContext is executed.